kolibri/plugins/document_pdf_render/assets/src/views/index.vue

```diff
@@ -39,40 +95,277 @@
+          if (
+            // Do not try to show the page if it is already renderered,
+            // already loading, or already rendering.
+            !this.pdfPages[pageNum].rendered &&
+            !this.pdfPages[pageNum].loading &&
+            !this.pdfPages[pageNum].rendering
+          ) {
+            this.pdfPages[pageNum].loading = true;
+            // If we already have a reference to the PDFJS page object, then use it,
+            // rather than refetching it.
+            if (!this.pdfPages[pageNum].pdfPage) {
+              this.pdfPages[pageNum].renderPromise = this.getPage(pageNum).then(this.startRender);
+            } else {
+              this.pdfPages[pageNum].renderPromise = this.startRender(this.pdfPages[pageNum].pdfPage);
+            }
           }
-          this.isFullscreen = true;
         }
       },
+      hidePage(pageNum) {
+        if (pageNum <= this.totalPages && pageNum > 0) {
+          // Only try to hide possibly existing pages.
+          if (!this.pdfPages[pageNum]) {
+            // No page to render, so do nothing
+            return;
+          }
+          const pdfPage = this.pdfPages[pageNum];
+          if (pdfPage.rendered) {
+            pdfPage.rendered = false;
+            // Already rendered, just remove canvas from DOM.
+            if (pdfPage.canvas) {
+              pdfPage.canvas.remove();
+            }
+          } else if (pdfPage.loading) {
+            // Otherwise, currently loading - let it finish the render promise,
+            // where the page is still being fetched
+            // then cancel the resulting renderTask.
+            const renderTask = pdfPage.renderTask;
+            pdfPage.renderPromise.then(() => {
+              renderTask && renderTask.cancel();
+            });
+          } else if (pdfPage.rendering) {
+            // Currently rendering, cancel the task directly
+            pdfPage.renderTask && pdfPage.renderTask.cancel();
+          }
+          // Clean everything up (destroys the pdf page object).
+          pdfPage.pdfPage && pdfPage.pdfPage.cleanup();
+          // Delete the reference so that this page is now a blank slate.
+          delete this.pdfPages[pageNum];
+        }
+      },
+      pageRef(index) {
+        return `pdfPage-${index}`;
+      },
+      // debouncing so we're not de/re-render many pages unnecessarily
+      checkPages: debounce(function() {
+        // Calculate the position of the visible top and the bottom of the pdfContainer
+        const top = this.$refs.pdfContainer.scrollTop;
+        const bottom = top + this.$refs.pdfContainer.clientHeight;
+        // Then work out which pages are visible to the user as a consequence
+        const topPageNum = Math.ceil(top / this.pageHeight);
+        const bottomPageNum = Math.ceil(bottom / this.pageHeight);
+
+        // Loop through all pages, show ones that are in the display window, hide ones that aren't
+        for (let i = 1; i <= this.totalPages; i++) {
```

Comment:
In this instance, I think a numeric loop is more appropriate, as we are looping over page numbers - to do a forEach would require creating a range Array which is clever, but not very readable:
`[...Array(this.totalPages).keys()].forEach(i => {`